src/biz/workspace/page_view.rs

```diff
@@ -0,0 +1,229 @@
+      get_page_collab_data_for_document(
+        collab_access_control_storage.clone(),
+        uid,
+        workspace_id,
+        view_id,
+      )
+      .await
+    },
+    collab_folder::ViewLayout::Grid
+    | collab_folder::ViewLayout::Board
+    | collab_folder::ViewLayout::Calendar => {
+      get_page_collab_data_for_database(
+        pg_pool,
+        collab_access_control_storage.clone(),
+        uid,
+        workspace_id,
+        view_id,
+      )
+      .await
+    },
+    collab_folder::ViewLayout::Chat => Err(AppResponseError::new(
+      ErrorCode::InvalidRequest,
+      "Page view for AI chat is not supported at the moment",
+    )),
+  }?;
+
+  let page_collab = PageCollab {
+    view: folder_view,
+    data: page_collab_data,
+    owner,
+    last_editor,
+  };
+
+  Ok(page_collab)
+}
+
+async fn get_page_collab_data_for_database(
+  pg_pool: &PgPool,
+  collab_access_control_storage: Arc<CollabAccessControlStorage>,
+  uid: i64,
+  workspace_id: Uuid,
+  view_id: &str,
+) -> Result<PageCollabData, AppResponseError> {
+  let ws_db_oid = select_workspace_database_oid(pg_pool, &workspace_id).await?;
+  let ws_db = get_latest_collab_encoded(
+    collab_access_control_storage.clone(),
+    GetCollabOrigin::User { uid },
+    &workspace_id.to_string(),
+    &ws_db_oid,
+    CollabType::WorkspaceDatabase,
+  )
+  .await?;
+  let mut ws_db_collab = collab_from_doc_state(ws_db.doc_state.to_vec(), &ws_db_oid)?;
+  let ws_db_body = WorkspaceDatabaseBody::open(&mut ws_db_collab);
+  let txn = ws_db_collab.transact();
+  let db_oid = ws_db_body
+    .get_database_meta_with_view_id(&txn, view_id)
+    .ok_or(AppResponseError::new(
+      ErrorCode::NoRequiredData,
+      format!("Database view {} not found", view_id),
+    ))?
+    .database_id;
+  let db = get_latest_collab_encoded(
+    collab_access_control_storage.clone(),
+    GetCollabOrigin::User { uid },
+    &workspace_id.to_string(),
+    &db_oid,
+    CollabType::Database,
+  )
+  .await?;
+  let db_collab = Collab::new_with_source(
+    CollabOrigin::Server,
+    &db_oid,
+    DataSource::DocStateV1(db.doc_state.to_vec()),
+    vec![],
+    false,
+  )
+  .map_err(|err| {
+    AppResponseError::new(
+      ErrorCode::Internal,
+      format!(
+        "Unable to create collab from object id {}: {}",
+        &db_oid, err
+      ),
+    )
+  })?;
+  let db_body = DatabaseBody::from_collab(&db_collab).unwrap();
+  let txn = db_collab.transact();
+  let inline_view_id = db_body.get_inline_view_id(&txn);
+  let row_ids: Vec<RowId> = db_body
+    .views
+    .get_row_orders(&txn, &inline_view_id)
+    .iter()
+    .map(|ro| ro.id.clone())
+    .collect();
+  let queries: Vec<QueryCollab> = row_ids
+    .iter()
+    .map(|row_id| QueryCollab {
+      object_id: row_id.to_string(),
+      collab_type: CollabType::DatabaseRow,
+    })
+    .collect();
+  let row_query_collab_results = collab_access_control_storage
+    .batch_get_collab(&uid, queries)
```

Comment:
I am thinking about refactor the `batch_get_collab` that returns a collab stream instead.